examples/common/pigweed/rpc_services/AttributeAccessorRegistry.h

```diff
@@ -0,0 +1,188 @@
+            if (perEndpointClusterAccessors.find(endpointClusterPair) == perEndpointClusterAccessors.end())
+            {
+                ChipLogError(Support, "Unregistration failed. No registration found for endpointId: %u , clusterId: %u",
+                             endpointId.Value(), clusterId);
+                return;
+            }
+            perEndpointClusterAccessors.erase(endpointClusterPair);
+        }
+        else
+        {
+            if (perClusterAccessors.find(clusterId) == perClusterAccessors.end())
+            {
+                ChipLogError(Support, "Unregistration failed. No registration found for clusterId: %u", clusterId);
+                return;
+            }
+            perClusterAccessors.erase(clusterId);
+        }
+    }
+
+    /**
+     * Unregister all attribute accessors that match this given endpoint.
+     * Cluster level attribute accessors do not get unregistered.
+     */
+    void UnregisterAllForEndpoint(EndpointId endpointId)
+    {
+        std::vector<std::pair<EndpointId, ClusterId>> removeRegistrations;
+        for (std::map<std::pair<EndpointId, ClusterId>, AttributeAccessor *>::iterator it = perEndpointClusterAccessors.begin();
+             it != perEndpointClusterAccessors.end(); ++it)
+        {
+            if ((it->first).first == endpointId)
+            {
+                removeRegistrations.push_back(it->first);
+            }
+        }
+        for (std::pair<EndpointId, ClusterId> endpointClusterPair : removeRegistrations)
+        {
+            ChipLogProgress(Support, "Removing attribute accessor registration for endpointId: %u , clusterId: %u",
+                            endpointClusterPair.first, endpointClusterPair.second);
+            perEndpointClusterAccessors.erase(endpointClusterPair);
+        }
+    }
+
+    /**
+     * Unregisters all attribute accessors.
+     */
+    void UnregisterAll()
+    {
+        perClusterAccessors.clear();
+        perEndpointClusterAccessors.clear();
+    }
+
+    /**
+     *  Get the registered attribute access override that is cluster-endpoint specific.
+     */
+    AttributeAccessor * Get(EndpointId aEndpointId, ClusterId aClusterId)
+    {
+        std::pair<EndpointId, ClusterId> endpointClusterPair = std::make_pair(aEndpointId, aClusterId);
+        if (perEndpointClusterAccessors.find(endpointClusterPair) != perEndpointClusterAccessors.end())
+        {
+            return perEndpointClusterAccessors.at(endpointClusterPair);
+        }
+        return nullptr;
+    }
+
+    /**
+     * Get the registered attribute access override that is cluster specific (Global to all endpoints)
+     */
+    AttributeAccessor * Get(ClusterId aClusterId)
+    {
+        if (perClusterAccessors.find(aClusterId) != perClusterAccessors.end())
+        {
+            return perClusterAccessors.at(aClusterId);
+        }
+        return nullptr;
+    }
```

Comment:
Not sure why we have to register per cluster and and per endpoint. In fact, why do we need to map, rather than simply allow iterating over all of the registered handlers, which could handle each more than 1 endpoint/cluster? The read/write receive a path, so it's possible to validate the endpoint/cluster in there. The model here adds complexity/separation that is not needed.